### DISCUSSION TITLE
Need some pointers using MoorIsolate

### DISCUSSION THREAD

**Discussion by frankvollebregt** [simolus3/drift#1407] [Q&A] [answered]
Hello, I'm trying to use isolates in my app to use the database, but something is going wrong.

For a bit of context, the app uses a Moor database, from which it streams several values into the front-end (`watch` with `StreamBuilder`s). Now there's two issues. Firstly, the `CircularProgressIndicator` starts to lag heavily when Moor is busy clearing/inserting/updating the database (quite a bit of data). As a result, I want to use a separate isolate to perform these database operations. Secondly (at a later point) I want the app to be able to load data from the server in the background, and store it in the database. In the page about isolates in [the documentation](https://moor.simonbinder.eu/docs/advanced-features/isolates/) a lot of info is mentioned, but I somehow can't get it to work the way I want it to.

Both when loading from my main isolate or from another isolate I spawned using the `FlutterIsolate` package, when I execute the code below no errors/exceptions occur, but the database seems to be empty.

```dart
// function from the documentation
DatabaseConnection _backgroundConnection() {
  // construct the database. You can also wrap the VmDatabase in a "LazyDatabase" if you need to run
  // work before the database opens.
  final database = VmDatabase.memory();
  return DatabaseConnection.fromExecutor(database);
}

doThingsInIsolate() async {
    // create the isolate with the database
    MoorIsolate isolate = await MoorIsolate.spawn(_backgroundConnection);
    DatabaseConnection connection = await isolate.connect();
    final MyDatabase db = MyDatabase.connect(connection);
    print('database is $db');    // instance of 'MyDatabase'
    print('the items: ${await db.allItems().get()}');    // expected to return [Item<>], but returns []

    // the actual DB stuff I want to do goes here
    // intermediate prints happen fine, but the database remains untouched
    await doActualDatabaseInteractions();

    db.close();
    isolate.shutdownAll();
}
```

If there's anything clearly wrong, I would love to know. If there is some working example code I can take a look at, which shows how to work with these, that would also be very helpful.

**Comment by simolus3** [comment, marked as answer]
Are you inserting data in your database's migration? If you're using a `VmDatabase.memory()` you'll get a clean in-memory database for each `MoorIsolate.spawn`, it would not get persisted anywhere. So it's not entirely unexpected that `allItems()` would return an empty list after opening a fresh database. You can re-use the existing `VmDatabase` you were using before isolates in `_backgroundConnection()`. When using things like `path_provider`, you'll also need to call `FlutterWidgetsBinding.ensureInitialized()` in the background isolate to setup platform channels.

Apart from that, the code looks correct to me though.

**Reply by frankvollebregt** [reply]
Ahh, I didn't even realise the VMDatabase is _meant_ to be empty originally. Of course... I will indeed need the path provider, but the documentation should be able to help me there :)

Thank you, I will mark this question as answered once I get it working (which is hopefully quickly)

**Reply by simolus3** [reply]
Good point, I'll add a comment to the documentation pointing out that a real database should probably be used and mention the `FlutterWidgetsBinding` thing. Let me know if you run into more problems with this :)

**Reply by davidmartos96** [reply]
@simolus3 Is it a new thing that the isolates work with Platform channels? As far as I know that is still a work in progress in the Flutter team

**Reply by simolus3** [reply]
Oh good point actually. If I remember correctly `path_provider` worked on a background isolate, but the docs also mention the other approach of computing the right path on the main isolate and sending that.

**Reply by frankvollebregt** [reply]
It works fine for me right now. I'm setting up an isolate for Moor in my `main.dart`, and when I want to do networking to retrieve new data, I create a new isolate using the `flutter_isolate` package (which does allow me to use other packages), and I use the `SendPort` to connect to the database isolate.

As an aside, I currently pass this `SendPort` down the widget tree to share it, but is there a recommended way to do this? Would the `provider` package be a good approach here?

**Reply by simolus3** [reply]
I don't have any preferences here, using `provider` sounds reasonable. I'd probably wrap the port then, though:

```dart
class MoorConnectPort {
  final SendPort connect;
  MoorConnectPort(this.connect);
}
```

Because if you just do `Provider.of<SendPort>(context)` it's not clear what kind of send port we're talking about :)

**Reply by frankvollebregt** [reply]
Thank you very much for all the help!
